### PR TITLE
feat: Hide UI elements through window postmessage

### PR DESCRIPTION
**Pull request by gabriellsh**
## Proposed changes (including videos or screenshots)

Made possible to hide some UI elements through the window.postMessage event.

Initial supported UI elements are:
Composer Toolbar
Room Toolbar
Message Toolbar
User actions toolbar

## Issue(s)

[CE-46](https://rocketchat.atlassian.net/browse/CE-46)
## Further comments

Example call to hide all UI elements
```
window.postMessage({
    event: 'overrideUi',
    hideActions: {
        roomToolbox:[
			'channel-settings',
			'team-info',
			'user-info-group',
			'user-info',
			'thread',
			'calls',
			'canned-responses',
			'clean-history',
			'contact-chat-history',
			'contact-profile',
			'discussions',
			'export-messages',
			'keyboard-shortcut-list',
			'members-list',
			'mentions',
			'otr',
			'pinned-messages',
			'push-notifications',
			'rocket-search',
			'room-info',
			'starred-messages',
			'start-call',
			'team-channels',
			'uploaded-files-list',
			'voip-room-info',
		],
		messageToolbox: [
			'reaction-message',
			'quote-message',
			'reply-in-thread',
			'forward-message',
			'reply-directly',
			'follow-message',
			'pin-message',
			'star-message',
			'permalink',
			'copy',
			'edit-message',
			'delete-message',
		],
		composerToolbox: [
			'video-message',
			'audio-message',
			'file-upload',
			'create-discussion',
			'webdav-add',
			'share-location'
		],
		userToolbox: [
			'openDirectMessage',
			'changeOwner',
			'changeLeader',
			'changeModerator',
			'openModerationConsole',
			'ignoreUser',
			'muteUser',
			'reportUser',
			'removeUser',
		],
    }
})
```

[CE-46]: https://rocketchat.atlassian.net/browse/CE-46?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apps/meteor/client/views/room/composer/messageBox/MessageBoxActionsToolbar/MessageBoxActionsToolbar.tsx [hunk 1]
```diff
@@ -1,54 +1,60 @@
...
 	rid: IRoom['_id'];
 	tmid?: IMessage['_id'];
-	isMicrophoneDenied?: boolean;
 };
 
-const MessageBoxActionsToolbar = ({
-	variant = 'large',
-	isRecording,
-	typing,
+const MessageBoxActionsToolbarBuilder = ({
```
Comment (tassoevan) [inline apps/meteor/client/views/room/composer/messageBox/MessageBoxActionsToolbar/MessageBoxActionsToolbar.tsx:19]:
```suggestion
const MessageBoxActionsToolbar = ({
```

File: apps/meteor/client/views/room/composer/messageBox/MessageBoxActionsToolbar/MessageBoxActionsToolbar.tsx [hunk 2]
```diff
@@ -1,54 +1,60 @@
...
-			{variant !== 'small' && actions}
-			{variant === 'small' && featuredAction}
-			<ActionsToolbarDropdown {...(variant === 'small' && { actions })} isRecording={isRecording} rid={rid} tmid={tmid} {...props} />
+			<MessageComposerActionsDivider />
+			{featured.map((action) => (
+				<MessageComposerAction
+					key={action.id}
+					{...action}
+					data-qa-id={action.id}
+					icon={action.icon as ComponentProps<typeof MessageComposerAction>['icon']}
```
Comment (tassoevan) [inline apps/meteor/client/views/room/composer/messageBox/MessageBoxActionsToolbar/MessageBoxActionsToolbar.tsx:52]: Why assert it?

File: apps/meteor/client/views/room/composer/messageBox/MessageBoxActionsToolbar/hooks/ToolbarAction.ts [hunk 3]
```diff
@@ -0,0 +1,8 @@
+export type ToolbarAction = {
+	title?: string;
+	disabled?: boolean;
+	onClick: (...params: any) => unknown;
+	icon: string;
```
Comment (tassoevan) [inline apps/meteor/client/views/room/composer/messageBox/MessageBoxActionsToolbar/hooks/ToolbarAction.ts:5]: Why not the union of icon names?

File: apps/meteor/client/views/room/composer/messageBox/MessageBoxActionsToolbar/hooks/useCreateDiscussionAction.tsx [hunk 4]
```diff
@@ -1,12 +1,16 @@
 import type { IRoom } from '@rocket.chat/core-typings';
 import { isRoomFederated } from '@rocket.chat/core-typings';
-import { Option, OptionIcon, OptionContent } from '@rocket.chat/fuselage';
 import { useTranslation, useSetting, usePermission, useSetModal } from '@rocket.chat/ui-contexts';
 import React from 'react';
 
 import CreateDiscussion from '../../../../../../components/CreateDiscussion';
+import type { ToolbarAction } from './ToolbarAction';
+
+export const useCreateDiscussionAction = (room?: IRoom): ToolbarAction => {
```
Comment (tassoevan) [inline apps/meteor/client/views/room/composer/messageBox/MessageBoxActionsToolbar/hooks/useCreateDiscussionAction.tsx:9]: Keep in mind the hooks can't be conditionally called, while components can.

**Review by tassoevan — CHANGES_REQUESTED** [review]

**Review by tassoevan — APPROVED** [review]

**Comment by MHFDoge** [comment]
Small question, Is there a reason this was introduced **strictly** for users who embed via iframe, and not just made to work with the existing role-based permission system? I have personally run into several issues over multiple years with things on my workspace that could be quickly solved if a variation of this feature was made available to those who use the standard web interface provided by the server application itself.

**Comment by gabriellsh** [comment]
@MHFDoge it's actually not exclusive to iframe, you can use it anywhere you can inject a custom script. This comes from common user requests that try to embed rocket.chat in another application, so it fits this need without being to complex. Coupling the UI to the role-based permission system is not something we want to do.

**Pull request merged**